File: benchmarks/js/tree.js
// benchmark: tree (JS: runs under both jsc and node)
// Build and traverse 30 full binary trees of depth 16; the engine's GC reclaims
// each tree (JS has no manual free). Checksum: node-visit count = 3932130.
"use strict";

function build(depth) {
  if (depth === 0) {
    return { left: null, right: null };
  }
  return { left: build(depth - 1), right: build(depth - 1) };
}

function check(node) {
  if (node.left === null) {
    return 1;
  }
  if (node.right === null) {
    return 1;
  }
  return 1 + check(node.left) + check(node.right);
}

function workload() {
  var DEPTH = 16, COUNT = 30;
  var checksum = 0;
  for (var i = 0; i < COUNT; i++) {
    var root = build(DEPTH);
    checksum += check(root);
    root = null;
  }
  return checksum;
}

var emit = (typeof print === "function") ? print : console.log;
function nowMs() { return performance.now(); }
(function () {
  var WARMUP = 3, TIMED = 11, checksum = 0, i;
  for (i = 0; i < WARMUP; i++) { checksum = workload(); }
  var times = new Array(TIMED);
  for (i = 0; i < TIMED; i++) {
    var t0 = nowMs();
    checksum = workload();
    var t1 = nowMs();
    times[i] = t1 - t0;
  }
  times.sort(function (a, b) { return a - b; });
  var median = times[(TIMED - 1) >> 1];
  emit(String(checksum) + " " + (median / 1000).toFixed(9));
})();

File: benchmarks/lua/tree.lua
-- benchmark: tree (LuaJIT)
-- Build and traverse 30 full binary trees of depth 16; LuaJIT's GC reclaims each
-- tree (Lua has no manual free). Checksum: node-visit count = 3932130.

local function build(depth)
  if depth == 0 then
    return {}
  end
  return { left = build(depth - 1), right = build(depth - 1) }
end

local function check(node)
  if node.left == nil then
    return 1
  end
  if node.right == nil then
    return 1
  end
  return 1 + check(node.left) + check(node.right)
end

local function workload()
  local DEPTH, COUNT = 16, 30
  local checksum = 0
  for _ = 1, COUNT do
    local root = build(DEPTH)
    checksum = checksum + check(root)
    root = nil
  end
  return checksum
end

local WARMUP, TIMED = 3, 11
local checksum = 0
for _ = 1, WARMUP do checksum = workload() end
local times = {}
for i = 1, TIMED do
  local t0 = os.clock()
  checksum = workload()
  local t1 = os.clock()
  times[i] = t1 - t0
end
table.sort(times)
local median = times[math.floor((TIMED + 1) / 2)]
io.write(string.format("%d %.9f\n", checksum, median))
